### PR TITLE
Add User model

### PR DESCRIPTION
**Pull request by sviatoslav-krupa**
Add User model by gem 'devise' with testing by rspec. Also, generate views and controllers for authentication and registration. All views with .html.slim extension.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class User < ApplicationRecord
+  enum role: %i[sale admin moderator]
```
Comment (f1nwe) [inline app/models/user.rb]: move it to const

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email,
+            presence: true,
+            length: { maximum: 255 },
+            format: { with: VALID_EMAIL_REGEX },
+            uniqueness: { case_sensitive: false }
+
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
```
Comment (f1nwe) [inline app/models/user.rb]: you don't need this all modules

File: app/views/devise/mailer/password_change.html.slim [hunk 3]
```diff
@@ -0,0 +1,6 @@
+p
+  | Hello
```
Comment (f1nwe) [inline app/views/devise/mailer/password_change.html.slim]: p "Hello #{@resource.email}" will look nicer.

File: app/views/devise/registrations/edit.html.slim [hunk 4]
```diff
@@ -0,0 +1,42 @@
+h2
+  | Edit
```
Comment (f1nwe) [inline app/views/devise/registrations/edit.html.slim]: Are you sure, that there will be ability for registration?

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+
+  it 'is valid with first name, last name, e-mail and password' do
+    user = User.new(
```
Comment (f1nwe) [inline spec/models/user_spec.rb]: generate test data with factorybot.
also read about let() and let()! and use it, not local variables for storing objects in tests.

File: test/fixtures/users.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (f1nwe) [inline test/fixtures/users.yml]: you don't need fixtures at all.
You should use factories.

File: test/models/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
```
Comment (f1nwe) [inline test/models/user_test.rb]: use rspec for tests.
remove test folder.
All test should be stored in /spec folder.

File: app/views/devise/registrations/edit.html.slim (hunk 4 above)
Comment (egorodessa) [reply app/views/devise/registrations/edit.html.slim]: yes. user as tenant will possible to register as i understood.

File: spec/factories/user.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :user, class: User do
+    first_name { Faker::Name.last_name }
+    last_name { Faker::Name.last_name }
+    email { Faker::Internet.safe_email }
```
Comment (romaslmd) [inline spec/factories/user.rb]: doesn't look good,
what about some uniqueness?

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  it 'is valid with first name, last name, e-mail and password' do
+    user = FactoryBot.build(:user)
+    expect(user).to be_valid
+  end
+
+  it 'is invalid without first name' do
+    user = FactoryBot.build(:user, first_name: nil)
+    user.valid?
+    expect(user.errors[:first_name]).to include("can't be blank")
```
Comment (romaslmd) [inline spec/models/user_spec.rb]: use shoulda matchers

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  it 'is invalid without last name' do
+    user = FactoryBot.build(:user, last_name: nil)
+    user.valid?
+    expect(user.errors[:last_name]).to include("can't be blank")
```
Comment (romaslmd) [inline spec/models/user_spec.rb]: shoulda

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  it 'is invalid without e-mail' do
+    user = FactoryBot.build(:user, email: nil)
+    user.valid?
+    expect(user.errors[:email]).to include("can't be blank")
```
Comment (romaslmd) [inline spec/models/user_spec.rb]: shoulda

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  it 'is invalid with duplicate e-mail' do
+    first_user = FactoryBot.build(:user, email: 'example@gmail.com')
+    first_user.save
+    second_user = FactoryBot.build(:user, email: 'example@gmail.com')
+    second_user.valid?
+    expect(second_user.errors[:email]).to include('has already been taken')
```
Comment (romaslmd) [inline spec/models/user_spec.rb]: and shoulda

File: app/models/user.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+class User < ApplicationRecord
+  ROLES = %i[sale admin moderator].freeze
```
Comment (romaslmd) [inline app/models/user.rb]: better to use hash-like
easier to manage roles and seek role by int identifier

File: spec/factories/user.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+FactoryBot.define do
+  sequence :email do |n|
```
Comment (romaslmd) [inline spec/factories/user.rb]: u could simple use sequence within factory
e.g.:

```
factory :user do
  sequence { |n| Faker::Internet.email.gsub('@', "_#{n}@") }
  # other attributes ..
end
```

File: spec/models/user_spec.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'validations' do
+    it { should validate_presence_of(:first_name) }
+    it { should validate_presence_of(:last_name) }
+    it { should validate_presence_of(:email) }
+    it { should validate_uniqueness_of(:email).case_insensitive }
```
Comment (romaslmd) [inline spec/models/user_spec.rb]: could you please cover all validations? length, format, [enums](https://github.com/thoughtbot/shoulda-matchers/blob/master/lib/shoulda/matchers/active_record/define_enum_for_matcher.rb) etc.

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 16]
```diff
@@ -0,0 +1,30 @@
+module Users
+  class OmniauthCallbacksController < Devise::OmniauthCallbacksController
```
Comment (tinados) [inline app/controllers/users/omniauth_callbacks_controller.rb]: Do we really need all these generated controllers? As I remember if you do not make any changes to it - you could drop them.
And by the way here is the inconsistency with options you declare in user model.

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    it { should define_enum_for(:role) }
+    it { should validate_length_of(:first_name).is_at_least(3) }
```
Comment (romaslmd) [inline spec/models/user_spec.rb]: check [this](https://github.com/thoughtbot/shoulda-matchers/blob/master/lib/shoulda/matchers/active_model/validate_length_of_matcher.rb#L121) out

File: app/views/devise/confirmations/new.html.slim [hunk 18]
```diff
@@ -0,0 +1,11 @@
+h2
+  | Resend confirmation instructions
```
Comment (Kmalynss) [inline app/views/devise/confirmations/new.html.slim]: And what about this views? Are they needed here?

File: spec/rails_helper.rb [hunk 19]
```diff
@@ -36,50 +18,22 @@
...
-    DatabaseCleaner.clean_with(:truncation)
-    DatabaseCleaner.strategy = :transaction
-  end
-
-  # start the transaction strategy as examples are run
-  config.around(:each) do |example|
-    DatabaseCleaner.cleaning do
-      example.run
+    FactoryBot.reload
+    config.include FactoryBot::Syntax::Methods
```
Comment (romaslmd) [inline spec/rails_helper.rb]: looks like this shouldn't be here

File: app/models/user.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
+class User < ApplicationRecord
+  ROLES = { sale: 0, admin: 1, moderator: 2 }.freeze
+  enum role: ROLES
+  before_save { email.downcase! }
+  validates :first_name, length: { minimum: 3, maximum: 50 }, presence: true
+  validates :last_name,  length: { minimum: 3, maximum: 50 }, presence: true
+  validates :password,   length: { minimum: 6, maximum: 25 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, length: { maximum: 255 },
```
Comment (skcc321) [inline app/models/user.rb]: why do you need this if you have already devise gem
please look at it's sources (validatable module) https://github.com/plataformatec/devise/blob/master/lib/devise/models/validatable.rb#L29-L36

File: app/models/user.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+class User < ApplicationRecord
+  ROLES = { sale: 0, admin: 1, moderator: 2 }.freeze
+  enum role: ROLES
+  before_save { email.downcase! }
+  validates :first_name, length: { minimum: 3, maximum: 50 }, presence: true
+  validates :last_name,  length: { minimum: 3, maximum: 50 }, presence: true
+  validates :password,   length: { minimum: 6, maximum: 25 }
```
Comment (skcc321) [inline app/models/user.rb]: devise already has such validation

File: spec/models/user_spec.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'validations' do
```
Comment (skcc321) [inline spec/models/user_spec.rb]: do not test validations. That's bullshit.
Test only valid factory.

File: spec/rails_helper.rb (hunk 19 above)
Comment (skcc321) [reply spec/rails_helper.rb]: I believe config is related to upper block

File: spec/rails_helper.rb [hunk 23]
```diff
@@ -36,50 +18,22 @@
...
-  #
-  # The different available types are documented in the features, such as in
-  # https://relishapp.com/rspec/rspec-rails/docs
-  config.infer_spec_type_from_file_location!
 
-  # Filter lines from Rails gems in backtraces.
-  config.filter_rails_from_backtrace!
-  # arbitrary gems may also be filtered via:
-  # config.filter_gems_from_backtrace("gem name")
+    config.fixture_path = "#{::Rails.root}/spec/fixtures"
```
Comment (skcc321) [inline spec/rails_helper.rb]: where are your factories? in spec/factories ?
if so you do not need that line

**Pull request merged**